xline/src/storage/execute_error.rs

```diff
@@ -1,134 +1,103 @@
+    UserAlreadyHasRole(String, String),
+    /// Password was given for no password user
+    #[error("password was given for no password user")]
+    NoPasswordUser,
     /// Role not found
-    pub(crate) fn role_not_found(rolename: &str) -> Self {
-        Self::AuthError(format!("role {rolename} not found"))
-    }
-
+    #[error("role {0} not found")]
+    RoleNotFound(String),
     /// Role already exists
-    pub(crate) fn role_already_exists(rolename: &str) -> Self {
-        Self::AuthError(format!("role {rolename} already exists"))
-    }
-
+    #[error("role {0} already exists")]
+    RoleAlreadyExists(String),
     /// Role not granted
-    pub(crate) fn role_not_granted(rolename: &str) -> Self {
-        Self::AuthError(format!("role {rolename} is not granted to the user"))
-    }
+    #[error("role {0} is not granted to the user")]
+    RoleNotGranted(String),
     /// Root role not exist
-    pub(crate) fn root_role_not_exist() -> Self {
-        Self::AuthError("root user does not have root role".to_owned())
-    }
-
+    #[error("root user does not have root role")]
+    RootRoleNotExist,
     /// Permission not granted
-    pub(crate) fn permission_not_granted() -> Self {
-        Self::AuthError("permission not granted to the role".to_owned())
-    }
-
+    #[error("permission not granted to the role")]
+    PermissionNotGranted,
     /// Permission not given
-    pub(crate) fn permission_not_given() -> Self {
-        Self::AuthError("permission not given".to_owned())
-    }
-
+    #[error("permission not given")]
+    PermissionNotGiven,
     /// Invalid auth management
-    pub(crate) fn invalid_auth_management() -> Self {
-        Self::AuthError("invalid auth management".to_owned())
-    }
-
+    #[error("invalid auth management")]
+    InvalidAuthManagement,
     /// Invalid auth token
-    pub(crate) fn invalid_auth_token() -> Self {
-        Self::AuthError("invalid auth token".to_owned())
-    }
-
+    #[error("invalid auth token")]
+    InvalidAuthToken,
     /// Token manager is not initialized
-    pub(crate) fn token_manager_not_init() -> Self {
-        Self::AuthError("token manager is not initialized".to_owned())
-    }
-
+    #[error("token manager is not initialized")]
+    TokenManagerNotInit,
     /// Token is not provided
-    pub(crate) fn token_not_provided() -> Self {
-        Self::AuthError("token is not provided".to_owned())
-    }
-
+    #[error("token is not provided")]
+    TokenNotProvided,
     /// Token is expired
-    pub(crate) fn token_old_revision(claim_rev: i64, cur_rev: i64) -> Self {
-        Self::AuthError(format!(
-            "token's revision {claim_rev} is older than current revision {cur_rev}"
-        ))
+    #[error("token's revision {0} is older than current revision {1}")]
+    TokenOldRevision(i64, i64),
+
+    /// Db error
+    #[error("db error: {0}")]
+    DbError(String),
+
+    /// Permission denied Error
+    #[error("permission denied")]
```

Comment:
It looks like there is nothing different between the consequences of `PermissionDenied`、`PermissionNotGiven` and `PermissionNotGranted`. In these scenarios, users are unable to perform their operations. We can keep the `PermissionDenied` error and discard the other two.

Reply:
The PermissionNotGiven is indeed redundant, it's should be a invalid reuqest error, I will fix it in #394. 

The PermissionDenied and PermissionNotGranted actually have different meanings. 
`PermissionDenied` means the user doesn't have the admin permission in one of its roles.
`PermissionNotGranted` means a role does not have a permission to a specific key.